service/history/replication/ack_manager.go

```diff
@@ -439,27 +439,55 @@ func (p *ackMgrImpl) generateHistoryReplicationTask(
 				return nil, err
 			}
 
-			// BranchToken will not set in get dlq replication message request
-			if len(taskInfo.BranchToken) == 0 {
-				taskInfo.BranchToken = branchToken
-			}
-
 			eventsBlob, err := p.getEventsBlob(
 				ctx,
-				taskInfo.BranchToken,
+				branchToken,
 				taskInfo.FirstEventID,
 				taskInfo.NextEventID,
 			)
 			if err != nil {
 				return nil, err
 			}
 
+			var newRunBranchToken []byte
+			if len(taskInfo.NewRunID) > 0 {
+				newRunContext, releaseFn, err := p.workflowCache.GetOrCreateWorkflowExecution(
```

Comment:
if new run ID is set, then here is taking another lock.
can we modify the logic to lock & get the branch token one by one, instead of double lock & double unlock?

Reply:
I see the concern here. Updated to not have the overlap.